### ISSUE TITLE
[Micropython] Crash with squares.py

### ISSUE THREAD

**Issue by RedGl0w** [numworks/epsilon#1665]
#### Describe the bug
When running squares.py on epsilon master, micropython crashes

#### ASAN trace
```
=================================================================
==17606==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffd512c48c0 at pc 0x5580d76935f0 bp 0x7ffd512c4880 sp 0x7ffd512c4870
READ of size 8 at 0x7ffd512c48c0 thread T0
    #0 0x5580d76935ef in gc_collect_root (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x64e5ef)
    #1 0x5580d76831be in gc_collect_regs_and_stack() (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x63e1be)
    #2 0x5580d7683231 in gc_collect (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x63e231)
    #3 0x5580d7693e84 in gc_alloc (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x64ee84)
    #4 0x5580d76964ef in m_malloc (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x6514ef)
    #5 0x5580d76a63a7 in mp_obj_new_float (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x6613a7)
    #6 0x5580d76a6b67 in mp_obj_float_binary_op (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x661b67)
    #7 0x5580d76c3c34 in mp_binary_op (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x67ec34)
    #8 0x5580d76c9e39 in mp_execute_bytecode (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x684e39)
    #9 0x5580d76a6ff6 in fun_bc_call (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x661ff6)
    #10 0x5580d76c9462 in mp_execute_bytecode (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x684462)
    #11 0x5580d76a6ff6 in fun_bc_call (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x661ff6)
    #12 0x5580d768271b in MicroPython::ExecutionEnvironment::runCode(char const*) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x63d71b)
    #13 0x5580d7209705 in Code::ConsoleController::runAndPrintForCommand(char const*) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1c4705)
    #14 0x5580d720977f in Code::ConsoleController::textFieldDidFinishEditing(TextField*, char const*, Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1c477f)
    #15 0x5580d731e8ac in TextField::privateHandleEvent(Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2d98ac)
    #16 0x5580d731ef27 in TextField::handleEvent(Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2d9f27)
    #17 0x5580d72eac65 in App::processEvent(Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2a5c65)
    #18 0x5580d72f021c in Container::dispatchEvent(Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2ab21c)
    #19 0x5580d71ecb7a in AppsContainer::dispatchEvent(Ion::Events::Event) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1a7b7a)
    #20 0x5580d73054ba in RunLoop::step() (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2c04ba)
    #21 0x5580d7305567 in RunLoop::runWhile(bool (*)(void*), void*) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x2c0567)
    #22 0x5580d71ed0d6 in AppsContainer::run() (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1a80d6)
    #23 0x5580d723dd71 in ion_main(int, char const* const*) (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1f8d71)
    #24 0x5580d71e9c66 in main (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1a4c66)
    #25 0x7ff56678d151 in __libc_start_main (/usr/lib/libc.so.6+0x28151)
    #26 0x5580d71eb74d in _start (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x1a674d)

Address 0x7ffd512c48c0 is located in stack of thread T0 at offset 0 in frame
    #0 0x5580d76830ee in gc_collect_regs_and_stack() (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x63e0ee)

  This frame has 1 object(s):
    [48, 248) 'regs' (line 253)
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-underflow (/home/redgl0w/Bureau/Sleep/output/release/simulator/linux/epsilon.bin+0x64e5ef) in gc_collect_root
Shadow bytes around the buggy address:
  0x10002a2508c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a2508d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a2508e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a2508f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a250900: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x10002a250910: 00 00 00 00 00 00 00 00[f1]f1 f1 f1 f1 f1 00 00
  0x10002a250920: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a250930: 00 00 00 00 00 00 00 f3 f3 f3 f3 f3 f3 f3 f3 f3
  0x10002a250940: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x10002a250950: 00 00 00 00 00 00 f1 f1 f1 f1 00 f2 f2 f2 00 f2
  0x10002a250960: f2 f2 00 00 f3 f3 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==17606==ABORTING
```

#### To Reproduce
1. Go to python
2. Start squares.py
3. type `squares()`

#### Expected behavior
Not to crash, obviously

#### Environment
 - Epsilon master
 - The platform(s) on which the problem happens: simulator (linux), and maybe on some device (investigating)

**Comment by boricj** [comment]
There's something definitively wrong somewhere inside MicroPython. I have frequent crashes with non-trivial scripts that run for a while. I've reduced it to the following script which, when executed through the "Execute script" option, will crash the device:
```
from math import *
while True:
  sin(0)
```

It doesn't crash on the simulator, but valgrind complains quite a lot:
```
==22598== Memcheck, a memory error detector
==22598== Copyright (C) 2002-2017, and GNU GPL'd, by Julian Seward et al.
==22598== Using Valgrind-3.14.0 and LibVEX; rerun with -h for copyright info
==22598== Command: ./output/release/simulator/linux/epsilon.bin
==22598== 
==22598== Conditional jump or move depends on uninitialised value(s)
==22598==    at 0x34200F: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Conditional jump or move depends on uninitialised value(s)
==22598==    at 0x34201B: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Conditional jump or move depends on uninitialised value(s)
==22598==    at 0x342024: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Use of uninitialised value of size 8
==22598==    at 0x34203B: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Conditional jump or move depends on uninitialised value(s)
==22598==    at 0x34204E: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Use of uninitialised value of size 8
==22598==    at 0x342056: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355F44: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
==22598== Use of uninitialised value of size 8
==22598==    at 0x341EA9: gc_mark_subtree (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34205D: gc_collect_root (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33BFFE: gc_collect_regs_and_stack() (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33C01B: gc_collect (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3422FF: gc_alloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x34336F: m_malloc (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x3491EB: mp_obj_new_float (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x355CA0: mp_execute_bytecode (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x349743: fun_bc_call (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353B45: mp_parse_compile_execute (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x33D5E7: mp_builtin___import__ (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598==    by 0x353942: mp_import_name (in /home/jblbeurope/epsilon/output/release/simulator/linux/epsilon.bin)
==22598== 
...
```

**Comment by RedGl0w** [comment]
I couldn't manage to reproduce crash on device, but some people did

**Comment by EmilieNumworks** [comment]
There are two different bugs you refer to.

1. The one exposed by Valgrind or the AddressSanitizer is not really a bug but a consequence of how we collect root in MicroPython. To determine the Python stack position, we initialize a pointer referred as `stackTop` before starting executing Python, then right before collecting the roots, we collect the registers values in a buffer stored on the stack whose address is the stack end. Then, we collect roots by scanning between these pointers. We might read through objects that could be reinterpreted as pointers pointing to the Python heap. However, this is really unlikely and has the only consequence of preventing one object to be swept. From valgrind or the AddressSanitizer point of vue, we read forbidden/uninitalized memory: they for instance instrumented the memory to wrap the registers buffer with forbidden flags and we're reading through them... This issue is known and so far we just added a comment:
```
/* Memory error detectors might find an error here as they might split regs
 * and stack memory zones. */
```
I didn't really think of how we could tell Valgrind or the AddressSanitizer that this specific behaviour there is tolerated?

2. The second bug happening on the device might be one we just fixed: the initialization of the Python stack top was done too late - after that some Python code has already been executed. This has been fixed by @HugoNumworks and should be pushed soon.
